zcash_client_sqlite/src/wallet/transparent.rs

```diff
@@ -32,3 +51,711 @@ pub(crate) fn detect_spending_accounts<'a>(
+        let output = to_unspent_transparent_output(row)?;
+        if excluded.contains(output.outpoint()) {
+            continue;
+        }
+
+        utxos.push(output);
+    }
+
+    Ok(utxos)
+}
+
+/// Returns the unspent balance for each transparent address associated with the specified account,
+/// such that the block that included the transaction was mined at a height less than or equal to
+/// the provided `max_height`.
+pub(crate) fn get_transparent_address_balances<P: consensus::Parameters>(
+    conn: &rusqlite::Connection,
+    params: &P,
+    account: AccountId,
+    max_height: BlockHeight,
+) -> Result<HashMap<TransparentAddress, NonNegativeAmount>, SqliteClientError> {
+    let mut stmt_address_balances = conn.prepare(
+        "SELECT u.address, SUM(u.value_zat)
+         FROM transparent_received_outputs u
+         JOIN transactions t
+         ON t.id_tx = u.transaction_id
+         WHERE u.account_id = :account_id
+         AND t.mined_height <= :max_height
+         AND u.id NOT IN (
+            SELECT txo_spends.transparent_received_output_id
+            FROM transparent_received_output_spends txo_spends
+            JOIN transactions tx ON tx.id_tx = txo_spends.transaction_id
+            WHERE
+              tx.block IS NOT NULL -- the spending tx is mined
+              OR tx.expiry_height IS NULL -- the spending tx will not expire
+              OR tx.expiry_height > :max_height -- the spending tx is unexpired
+         )
+         GROUP BY u.address",
+    )?;
+
+    let mut res = HashMap::new();
+    let mut rows = stmt_address_balances.query(named_params![
+        ":account_id": account.0,
+        ":max_height": u32::from(max_height),
+    ])?;
+    while let Some(row) = rows.next()? {
+        let taddr_str: String = row.get(0)?;
+        let taddr = TransparentAddress::decode(params, &taddr_str)?;
+        let value = NonNegativeAmount::from_nonnegative_i64(row.get(1)?)?;
+
+        res.insert(taddr, value);
+    }
+
+    Ok(res)
+}
+
+pub(crate) fn add_transparent_account_balances(
+    conn: &rusqlite::Connection,
+    summary_height: BlockHeight,
+    account_balances: &mut HashMap<AccountId, AccountBalance>,
+) -> Result<(), SqliteClientError> {
+    let mut stmt_account_balances = conn.prepare(
+        "SELECT u.account_id, SUM(u.value_zat)
+         FROM transparent_received_outputs u
+         JOIN transactions t
+         ON t.id_tx = u.transaction_id
+         WHERE t.mined_height <= :summary_height
+         -- and the received txo is unspent
+         AND u.id NOT IN (
+           SELECT transparent_received_output_id
+           FROM transparent_received_output_spends txo_spends
+           JOIN transactions tx
+             ON tx.id_tx = txo_spends.transaction_id
+           WHERE tx.block IS NOT NULL -- the spending tx is mined
+           OR tx.expiry_height IS NULL -- the spending tx will not expire
+           OR tx.expiry_height > :summary_height -- the spending tx is unexpired
```

Comment:
It could be that the spending tx is unmined and will not expire, but is dependent on another transaction that is unmined and *will* expire.

This is a pre-existing issue. I don't think there's any way to fix it from this side of the lightwalletd protocol. A full node will keep only transactions that *could* be mined in its mempool (subject also to [ZIP 401](https://zips.z.cash/zip-0401) if the mempool becomes full), but a light client doesn't have enough information to do that.

Reply:
Filed as #1426.